starlette/routing.py

```diff
@@ -541,29 +541,47 @@ async def lifespan(self, scope: Scope, receive: Receive, send: Send) -> None:
         Handle ASGI lifespan messages, which allows us to manage application
         startup and shutdown events.
         """
-        first = True
         app = scope.get("app")
         await receive()
+
+        async def as_aiter(gen: typing.Generator) -> typing.AsyncGenerator:
+            for val in gen:
+                yield val
+
+        lifespan_gen = (
+            self.lifespan_context(app)
+            if inspect.isasyncgenfunction(self.lifespan_context)
+            else as_aiter(self.lifespan_context(app))  # type: ignore
+        )
+
+        await self._lifespan_iter(lifespan_gen, receive, send)
+
+    async def _lifespan_iter(
+        self, gen: typing.AsyncGenerator, receive: Receive, send: Send
+    ) -> None:
         try:
```

Comment:
Inspiration: https://github.com/agronholm/anyio/blob/3.2.0/src/anyio/pytest_plugin.py#L71-L84

Reply:
I think just making lifespan context an asynccontextmanager factory makes all this easier